File: micro/form_hashmap_false_sharing.py
"""Python equivalent of form_hashmap_false_sharing.hl.

Two threads hammer disjoint keys of a shared dict. The GIL
serializes Python bytecode execution, so threads don't run in
parallel — but a threading.Lock-protected dict is the closest
analogue of Hale's `sync = serialized` shape: explicit per-
map lock, sequential mutations. The bench measures dict-set +
lock-acquire cost; expected throughput is significantly lower
than Hale or Go because Python's per-op overhead is heavy
even before adding sync.
"""

import threading
import time


class Registry:
    __slots__ = ("lock", "entries")
    def __init__(self):
        self.lock = threading.Lock()
        self.entries = {}

    def set(self, id_, v):
        with self.lock:
            self.entries[id_] = (id_, v)

    def __len__(self):
        with self.lock:
            return len(self.entries)


per_writer = 100_000
total_expected = per_writer * 2
reg = Registry()

def writer_a():
    for i in range(per_writer):
        reg.set(i * 2, i)

def writer_b():
    for i in range(per_writer):
        reg.set(i * 2 + 1, i)

t_a = threading.Thread(target=writer_a)
t_b = threading.Thread(target=writer_b)

t0 = time.monotonic_ns()
t_a.start()
t_b.start()
t_a.join()
t_b.join()
elapsed = time.monotonic_ns() - t0

print(f"per_writer={per_writer}")
print(f"total={len(reg)}")
print(f"elapsed_ns={elapsed}")

File: micro/form_hashmap_false_sharing.go
// Go equivalent of form_hashmap_false_sharing.hl.
//
// Two goroutines hammer disjoint keys (even / odd ids) of a
// shared map. Cross-goroutine access via sync.Mutex (the
// canonical Go shape for a shared mutating map; sync.Map is
// optimized for "many readers, occasional writer" which isn't
// this workload). runtime.LockOSThread on each goroutine pins
// them to distinct OS threads for the duration — closest
// analogue of Hale's cooperative pools each owning their own
// OS thread.
//
// Lock discipline matches Hale's `sync = serialized`: every
// mutate takes the per-map lock. Throughput is bounded by
// contention; this is the safe-but-slow baseline both
// languages share.
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

type Counter struct {
	Id    int
	Value int
}

type Registry struct {
	mu      sync.Mutex
	entries map[int]Counter
}

func (r *Registry) Set(c Counter) {
	r.mu.Lock()
	r.entries[c.Id] = c
	r.mu.Unlock()
}

func (r *Registry) Len() int {
	r.mu.Lock()
	n := len(r.entries)
	r.mu.Unlock()
	return n
}

func main() {
	perWriter := 100000
	total := perWriter * 2
	reg := &Registry{entries: make(map[int]Counter)}

	var wg sync.WaitGroup
	wg.Add(2)

	t0 := time.Now()

	// Goroutine A — even ids
	go func() {
		runtime.LockOSThread()
		defer wg.Done()
		for i := 0; i < perWriter; i++ {
			reg.Set(Counter{Id: i * 2, Value: i})
		}
	}()

	// Goroutine B — odd ids
	go func() {
		runtime.LockOSThread()
		defer wg.Done()
		for i := 0; i < perWriter; i++ {
			reg.Set(Counter{Id: i*2 + 1, Value: i})
		}
	}()

	wg.Wait()
	elapsed := time.Since(t0).Nanoseconds()

	fmt.Printf("per_writer=%d\n", perWriter)
	fmt.Printf("total=%d\n", reg.Len())
	fmt.Printf("elapsed_ns=%d\n", elapsed)
	_ = total
}
